packages/flutter/lib/src/services/text_boundary.dart

```diff
@@ -177,3 +177,159 @@ class DocumentBoundary extends TextBoundary {
+    // Handles outside of right bound.
+    if (position.offset > _text.length || (position.offset == _text.length  && position.affinity == TextAffinity.downstream)) {
+      position = TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    // Handles outside of left bound.
+    if (position.offset <= 0) {
+      return const TextPosition(offset: 0);
+    }
+    int index = position.offset;
+    if (position.affinity == TextAffinity.downstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+      return position;
+    }
+
+    for (index -= 1; index >= 0; index -= 1) {
+      if (!TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+        return TextPosition(offset: index + 1, affinity: TextAffinity.upstream);
+      }
+    }
+    return const TextPosition(offset: 0);
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    // Handles outside of right bound.
+    if (position.offset >= _text.length) {
+      return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+    }
+    // Handles outside of left bound.
+    if (position.offset < 0 || (position.offset == 0 && position.affinity == TextAffinity.upstream)) {
+      position = const TextPosition(offset: 0);
+    }
+
+    int index = position.offset;
+    if (position.affinity == TextAffinity.upstream && !TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index - 1))) {
+      return position;
+    }
+
+    for (; index < _text.length; index += 1) {
+      if (!TextLayoutMetrics.isWhitespace(_text.codeUnitAt(index))) {
+        return TextPosition(offset: index);
+      }
+    }
+    return TextPosition(offset: _text.length, affinity: TextAffinity.upstream);
+  }
+}
+
+/// Gets the boundary by calling the [outer] and pipe the result to
+/// [inner].
+///
+/// This class is useful if one wants to ignore certain text before finding the
+/// text boundary. For example, use [WhitespaceBoundary] as the
+/// [outer] to ignores any white space before finding the boundary
+/// of [inner] if the input position happens to be a whitespace
+/// character.
+class ExpandedTextBoundary extends TextBoundary {
+  /// Creates a [ExpandedTextBoundary] with inner and outter boundaries
+  const ExpandedTextBoundary({required this.inner, required this.outer});
+
+  /// The inner boundary to call with the result from [outer].
+  final TextBoundary inner;
+
+  /// The outer boundary to call with the input position.
+  ///
+  /// The result is piped to the [inner] before returning to the caller.
+  final TextBoundary outer;
+
+  @override
+  TextPosition getLeadingTextBoundaryAt(TextPosition position) {
+    return inner.getLeadingTextBoundaryAt(
+      outer.getLeadingTextBoundaryAt(position),
+    );
+  }
+
+  @override
+  TextPosition getTrailingTextBoundaryAt(TextPosition position) {
+    return inner.getTrailingTextBoundaryAt(
+      outer.getTrailingTextBoundaryAt(position),
+    );
+  }
```

Comment:
Consider overridding 
```
TextRange getTextBoundaryAt(TextPosition position)
```
too.

This method exists because it's sometimes slightly cheaper to query both the leading + trailing than to query them separately. (for example,  the low-level line boundary call on `ui.Paragraph` actually gives you both offsets so there's no need to call that twice to get both the leading boundary and trailing boundary). That probably should have been documented...

Reply:
Ah `LineBreak` didn't override that. Could you add that?

Reply:
added line break and wordboundary override, but I can't find a meaningful implementation of ExpandedTextBoundary.getTextBoundaryAt since the return value doesnot contain affinity. I can't use the return value of the outer textboundary to call inner textboundary.

Reply:
That's sad. Can we assume the TextRange's start position is always downstream and end position upstream?

Reply:
Well if the TextRange is collapsed we'll lose information. Maybe we should make `TextPosition` a subclass of `TextRange`?

Reply:
> That's sad. Can we assume the TextRange's start position is always downstream and end position upstream?

That is not the case for whitespace text boundary.

`abc def` the position at the whitespace in the middle will return
leading = TextPosition(3, upstream)
trailing = TextPosition(4, downstream)

Reply:
I think TextRange.start and TextRange.end should be a TextPosition instead of int